packages/lesswrong/server/users/permanentDeletion.ts

```diff
@@ -0,0 +1,129 @@
+import { getAdminTeamAccount } from "../callbacks/commentCallbacks";
+import { loggerConstructor } from "@/lib/utils/logging";
+import { mailchimpAPIKeySetting } from "../serverSettings";
+import { mailchimpEAForumListIdSetting, mailchimpForumDigestListIdSetting } from "@/lib/publicSettings";
+import md5 from "md5";
+import { captureException } from "@sentry/core";
+import { auth0RemoveAssociationAndTryDeleteUser } from "../authentication/auth0";
+
+/**
+ * Permanently delete a user from a Mailchimp list. Note
+ * that this means they can't sign back up using the API, only with the mailchimp
+ * signup form, so this should only be used if a user requests to permanently
+ * delete their data.
+ *
+ * See here for discussion of the api/sign-up restriction: https://stackoverflow.com/questions/52198510/mailchimp-resubscribe-a-deleted-member-causes-the-api-to-return-a-400-bad-reques
+ */
+const permanentlyDeleteFromMailchimpList = async ({
+  listId,
+  emailHash,
+  user,
+  logger
+}: {
+  listId: string;
+  emailHash: string;
+  user: DbUser;
+  logger: ReturnType<typeof loggerConstructor>;
+}) => {
+  const mailchimpAPIKey = mailchimpAPIKeySetting.get();
+
+  try {
+    const response = await fetch(
+      `https://us8.api.mailchimp.com/3.0/lists/${listId}/members/${emailHash}/actions/delete-permanent`,
+      {
+        method: "POST",
+        headers: {
+          "Content-Type": "application/json",
+          Authorization: `API_KEY ${mailchimpAPIKey}`,
+        },
+      }
+    );
+
+    if (!response.ok) {
+      const json = await response.json();
+      if (json.status !== 404) { // 404 just means they weren't subscribed to begin with
+        // eslint-disable-next-line no-console
+        console.error(
+          `Failed to permanently delete user ${user.displayName} from Mailchimp list ${listId}. Response: `,
+          json
+        );
+      }
+    } else {
+      logger(`Permanently deleted user with display name "${user.displayName}" from Mailchimp list ${listId}`);
+    }
+  } catch (e) {
+    // eslint-disable-next-line no-console
+    console.error(e);
+    captureException(e);
+  }
+};
+
+async function permanentlyDeleteUser(user: DbUser) {
+  const logger = loggerConstructor(`permanentlyDeleteUsers`);
+  const adminContext = createAdminContext();
+  const adminTeamAccount = await getAdminTeamAccount();
+  if (!adminTeamAccount) throw new Error("Couldn't find admin team account");
+
+  // Precaution: Ensure the soft-deleting callbacks have run (unsubscribing from mailchimp, reindexing elasticsearch content)
+  await updateMutator({
+    collection: Users,
+    documentId: user._id,
+    set: { deleted: true },
+    validate: false,
+    context: adminContext,
+    currentUser: adminContext.currentUser,
+  })
+  // Wait until async callbacks finish. This is overcautious, as there should be no need for the callbacks to refetch the user object
+  await new Promise(resolve => setTimeout(resolve, 5000));
```

Comment:
In the shell script you can use a "wait for no pending callbacks". Obviously won't work when you go to the server.

Reply:
Yep the fact that this needs to run on the live server means we can't use that, and I think it's better to handle the two cases in the same way